### PR TITLE
Add PGBouncer installation databases

### PR DESCRIPTION
**Pull request by gabrieljackson**
This new database type pools database connections via the PGBouncer
utility running inside the cluster. Installations are assigned to
a logical database in the database cluster and are given a schema
in which to store their data. Authentication is performed through
an authorization query from a pgbouncer user to the backing database
to ensure the user credentials are valid.

This change also includes a large amount of refactoring in the AWS
package in an attempt to reuse existing logic.

Fixes https://mattermost.atlassian.net/browse/MM-34809 and https://mattermost.atlassian.net/browse/MM-34810

Reviewers Notes:
 - This is very light on unit tests. These will be added as soon as we validate that connection pooling is leading to the expected scaling increase in the test account.
 - Logical databases are currently hardcoded to 10 installation schemas. API endpoints will be added for updating this as soon our initial tests conclude with positive results.
 - There is a lot of additional refactoring that should be done in the aws package to simplify logic and remove duplicated code.
 - The mappings for schema > logical database are currently stored directly in the multitenant database object. This probably isn't the best longterm solution and will be remedied with a follow-up database migration.

Results from a full ctest database suite run:
```
INFO[2021-07-01T13:46:17-04:00] ====================================================
INFO[2021-07-01T13:46:17-04:00] TEST RESULTS
INFO[2021-07-01T13:46:17-04:00] ====================================================
INFO[2021-07-01T13:46:17-04:00] PASS: aws-multitenant-rds-postgres-pgbouncer (0.91 minutes)
INFO[2021-07-01T13:46:17-04:00] PASS: aws-multitenant-rds-postgres (0.67 minutes)
INFO[2021-07-01T13:46:17-04:00] PASS: aws-multitenant-rds (0.83 minutes)
INFO[2021-07-01T13:46:17-04:00] PASS: aws-rds-postgres (5.63 minutes)
INFO[2021-07-01T13:46:17-04:00] PASS: aws-rds (5.20 minutes)
INFO[2021-07-01T13:46:17-04:00] PASS: mysql-operator (2.30 minutes)
INFO[2021-07-01T13:46:17-04:00] ====================================================
```

Release Notes
```release-note
Add PGBouncer installation databases
```

### CODE REVIEW

**Review by Szymongib — COMMENTED** [review]
Great work! It was a big undertaking 🚀 

I think the database code could use some love in the future (maybe worth logging some tickets) 😄 

For now, I have one concern regarding the DB migration, we might need to adjust them slightly, besides that looks good.

**Comment by gabrieljackson** [comment]
> I think the database code could use some love in the future

Yep! Totally agree. If the PGBouncer database option works well then it will be a good time for some serious cleanup.

**Review by Szymongib — APPROVED** [review]

**Review by angeloskyratzakos — APPROVED** [review]
Amazing work Gabe! That was indeed a big undertaking as Szymon mentioned

File: internal/tools/aws/constants.go [hunk 1]
```diff
@@ -203,6 +203,15 @@ const (
 	// schemas allowed in a Posgres multitenant RDS database cluster.
 	DefaultRDSMultitenantDatabasePostgresCountLimit = 300
 
+	// DefaultRDSMultitenantPGBouncerDatabasePostgresCountLimit is the maximum
+	// number of schemas allowed in a Posgres multitenant RDS database cluster
```
Comment (stylianosrigas) [inline internal/tools/aws/constants.go:207]: @gabrieljackson Is this basically the max number of installations per db cluster?

Comment (stylianosrigas) [inline internal/tools/aws/constants.go:207]:
```suggestion
	// number of schemas allowed in a Postgres multitenant RDS database cluster
```

File: internal/tools/aws/constants.go [hunk 2]
```diff
@@ -203,6 +203,15 @@ const (
...
+	// with a PGBouncer proxy.
+	DefaultRDSMultitenantPGBouncerDatabasePostgresCountLimit = 1000
```
Comment (stylianosrigas) [inline internal/tools/aws/constants.go:209]: @gabrieljackson Is this basically the max number of installations per db cluster?
 Write

File: internal/tools/aws/constants.go (hunk 1 above)
Comment (gabrieljackson) [reply internal/tools/aws/constants.go:207]: Thanks! I will clean this up in a change I am working on now.

File: internal/tools/aws/constants.go (hunk 2 above)
Comment (gabrieljackson) [reply internal/tools/aws/constants.go:209]: Yeah, it's completely arbitrary and just a value to get us started.

**Review by stylianosrigas — APPROVED** [review]
Just a question, amazing work 🚀

**Pull request merged**